diskv.go

```diff
@@ -216,7 +216,16 @@ func (d *Diskv) createKeyFileWithLock(pathKey *PathKey) (*os.File, error) {
 		return f, nil
 	}
 
-	mode := os.O_WRONLY | os.O_CREATE | os.O_TRUNC // overwrite if exists
+	// We call os.Remove before the call to OpenFile because otherwise we can
+	// mess with existing readers. If someone calls ReadStream, then proceeds
+	// to read from it slowly, then someone else calls WriteStream, the reader
+	// will start to get the *new* resource contents part-way through.
+	// By calling os.Remove, we unlink the existing file, but since the process
+	// still has a handle on the file, the existing readers can continue to
+	// read from it. Meanwhile we're free to update what's actually on-disk.
+	os.Remove(d.completeFilename(pathKey))
+
+	mode := os.O_WRONLY | os.O_CREATE | os.O_TRUNC // overwrite if exists somehow
 	f, err := os.OpenFile(d.completeFilename(pathKey), mode, d.FilePerm)
```

Comment:
Hmm... what do you think about creating the new file as d.completeFilename(pathKey) plus some random extension, and then relying on the existing behavior in writeStreamWithLock to atomically move it to the proper filename? That would accomplish the same thing, I think, and avoid this intermediate phase...

Reply:
This brings up an excellent point: the TempDir flag already fixes this problem by doing basically that.

The problem is that if you instantiate a diskv *without* setting the TempDir option, you cannot safely use ReadStream, which seems bad.

What's your opinion on making createKeyFileWithLock *always* write to d.completeFilename(pathKey)+extension? We would leave TempDir in the Options struct, but note that it is deprecated since *all* writes are now atomic.

If you have many slow readers using ReadStream, coupled with frequent writes, you will end up consuming additional disk space because the OS is keeping the old files around until the readers complete... but the alternative is handing back invalid data to the readers, which seems much worse to me!

Reply:
> What's your opinion on making createKeyFileWithLock always write to d.completeFilename(pathKey)+extension?

Yeah, I think I like this better. (Assuming the +extension bit is the random extension I mentioned.)

> If you have many slow readers using ReadStream, coupled with frequent writes, you will end up consuming additional disk space because the OS is keeping the old files around until the readers complete... 

I think this is basically unavoidable. If it became a problem, you could maybe interrupt readers.

You OK with this change?